File: javascript/src/p0009.js
/**
 * Project Euler Problem 9
 *
 * Brute force on this problem has been very effective, so why stop now
 *
 * Problem:
 *
 * A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,
 * a**2 + b**2 = c**2
 *
 * For example, 3**2 + 4**2 = 9 + 16 = 25 = 5**2.
 *
 * There exists exactly one Pythagorean triplet for which a + b + c = 1000.
 * Find the product abc.
 *
 * @return {number}
 */
exports.p0009 = function() {
    for (let c = 3; ; c++) {
        const cSquared = c * c;
        for (let b = 2; b < c; b++) {
            const bSquared = b * b;
            for (let a = 1; a < b; a++) {
                const aSquared = a * a;
                if (aSquared + bSquared == cSquared && a + b + c === 1000) {
                    return a * b * c;
                }
            }
        }
    }
    return 0;
};

File: cplusplus/src/p0009.cpp
/*
Project Euler Problem 9

This was fairly short to code, but it took me a minute to figure out how to deal with the lack of multi-loop breaking

Problem:

A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,
a2 + b2 = c2

For example, 32 + 42 = 9 + 16 = 25 = 52.

There exists exactly one Pythagorean triplet for which a + b + c = 1000.
Find the product abc.
*/
#ifndef EULER_P0009
#define EULER_P0009
#include <stdint.h>
#include <iostream>

uint64_t p0009() {
    uint64_t answer = 0;
    for (uint32_t c = 3; !answer && c < 1000; c++) {
        for (uint32_t b = 2; b < c; b++) {
            uint32_t a = 1000 - c - b;
            if (a < b && a*a + b*b == c*c) {
                answer = (uint64_t) a * b * c;
                break;
            }
        }
    }
    return answer;
}

#ifndef UNITY_END
int main(int argc, char const *argv[]) {
    std::cout << p0009() << std::endl;
    return 0;
}
#endif
#endif
